### PR TITLE
8268361: Fix the infinite loop in next_line

### PR DESCRIPTION
**Pull request by UncleNine**
If the /proc/stat mount point is changed in container environment, the while loop may lead to 100% cpu usage.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268361](https://bugs.openjdk.java.net/browse/JDK-8268361): Fix the infinite loop in next_line

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to e4f90ebda68f576c35c1f24e35bb9cecf8c561da
 * [Severin Gehwolf](https://openjdk.java.net/census#sgehwolf) (@jerboaa - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4378/head:pull/4378` \
`$ git checkout pull/4378`

Update a local copy of the PR: \
`$ git checkout pull/4378` \
`$ git pull https://git.openjdk.java.net/jdk pull/4378/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4378`

View PR using the GUI difftool: \
`$ git pr show -t 4378`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4378.diff">https://git.openjdk.java.net/jdk/pull/4378.diff</a>

</details>

### CODE REVIEW

**Comment by kelthuzadx** [comment]
I have encountered a similar problem before, I've filed a new issue https://bugs.openjdk.java.net/browse/JDK-8268361 for this. You can change your title to "8268361: Fix the infinite loop in next_line", then OpenJDK bot would associate your PR with the corresponding JBS issue.

Best regards,
Yang

File: src/jdk.management/linux/native/libmanagement_ext/UnixOperatingSystem.c [hunk 1]
```diff
@@ -64,7 +64,10 @@ static struct perfbuf {
 #define NS_PER_SEC 1000000000
 
 static void next_line(FILE *f) {
-    while (fgetc(f) != '\n');
+    int c;
+    do {
+        c = fgetc(f);
+    } while(c != '\n' && c != EOF);
```
Comment (dholmes-ora) [inline src/jdk.management/linux/native/libmanagement_ext/UnixOperatingSystem.c:70]: Style nit: please add space before (

Comment (dholmes-ora) [reply src/jdk.management/linux/native/libmanagement_ext/UnixOperatingSystem.c:70]: It is not obvious to me that the caller of next_line will handle the fact that we have hit EOF?

**Review by dholmes-ora — COMMENTED** [review]
Hi,

This fixes the potential infinite loop, but is it sufficient?

Thanks,
David

**Comment by jerboaa** [comment]
> If the /proc/stat mount point is changed in container environment, the while loop may lead to 100% cpu usage.

In what way have you observed /proc/stat being changed which manifests in 100% cpu usage?

File: src/jdk.management/linux/native/libmanagement_ext/UnixOperatingSystem.c (hunk 1 above)
Comment (UncleNine) [reply src/jdk.management/linux/native/libmanagement_ext/UnixOperatingSystem.c]: In my case,  it happened in the container environment.
the /proc filesystem of the container is provided by  lxcfs, but a lxcfs bug may make the /proc/stat mount point change, then  the file descriptor is different and fgetc function returns an EOF on error,  But c != '\n' is true and it leads to the  infinit loop.
Below are our flamegraph in the production , it happends on serveral frameworks（micrometer, elasticsearch..）which use the api "sun/management/OperatingSystemImpl.getSystemCpuLoad"
107140246-bb490f80-695b-11eb-9f62-0116079f4f4a

**Comment by UncleNine** [comment]
> > If the /proc/stat mount point is changed in container environment, the while loop may lead to 100% cpu usage.
> 
> In what way have you observed /proc/stat being changed which manifests in 100% cpu usage?

In my case, it happened in the container environment.
the /proc filesystem of the container is provided by lxcfs, but a lxcfs bug may make the /proc/stat mount point change.
It was caused by a lxcfs-3.1.2 segment fault's exit

File: src/jdk.management/linux/native/libmanagement_ext/UnixOperatingSystem.c (hunk 1 above)
Comment (UncleNine) [reply src/jdk.management/linux/native/libmanagement_ext/UnixOperatingSystem.c]: I use  top -Hp $PID command and the thread show 100% cpu usage
top
jtack

Comment (jerboaa) [reply src/jdk.management/linux/native/libmanagement_ext/UnixOperatingSystem.c]: @UncleNine We should handle the case in the `get_totalticks()` function - which seems to be the only user of `next_line()` - when `next_line()` returns EOF as David said. One way would be to return the 'c' character read in `next_line` and if it's EOF, return -2 in `get_totalticks()`.

Comment (UncleNine) [reply src/jdk.management/linux/native/libmanagement_ext/UnixOperatingSystem.c]: all right, i have updated the code

**Review by jerboaa — COMMENTED** [review]
Looks fine to me. Please fix the jcheck (whitespace) issue.

**Review by dholmes-ora — APPROVED** [review]
This seems better to me now.

Thanks,
David

**Review by sspitsyn — APPROVED** [review]

**Review by jerboaa — APPROVED** [review]
LGTM. Thanks for your patience!

**Comment by dholmes-ora** [comment]
@UncleNine  please do not force push commits to an open PR as it makes it difficult for reviewers to track the changes. The PR can contain as many commits as you like as it will all be squashed to a single clean commit when integrating. Thanks.

**Comment by tstuefe** [comment]
Would it not be better to read the whole content of /proc/stat with a single read() call instead of line by line? I don't think proc fs guarantees any kind of consistency with separate reads.

**Comment by UncleNine** [comment]
> @UncleNine please do not force push commits to an open PR as it makes it difficult for reviewers to track the changes. The PR can contain as many commits as you like as it will all be squashed to a single clean commit when integrating. Thanks.

Sorry,

**Comment by UncleNine** [comment]
> @UncleNine please do not force push commits to an open PR as it makes it difficult for reviewers to track the changes. The PR can contain as many commits as you like as it will all be squashed to a single clean commit when integrating. Thanks.

Sorry,

**Comment by UncleNine** [comment]
> Would it not be better to read the whole content of /proc/stat with a single read() call instead of line by line? I don't think proc fs guarantees any kind of consistency with separate reads.

yes, so fgetc's return value should need to be checked

**Pull request closed without merging**